### DISCUSSION TITLE
Exception raised with duplicate keys using `x-for` directive

### DISCUSSION THREAD

**Discussion by csfrancis** [alpinejs/alpine#2830] [7. Bugs (archived)] [answered]
Please forgive me if I'm missing something, but I believe there's a bug when using `x-for` with duplicate keys. For example, using the following data:
```js
{
  trades: [
    {
      'user': 'foo',
      'time': '2022-04-05T12:30:07-0500'
    },
    {
      'user': 'bar',
      'time': '2022-04-05T12:30:07-0500'
    }
    ]
}
```

When using `trade.time` as the `:key` on an `x-for` directive, an exception is raised [here](https://github.com/alpinejs/alpine/blob/v3.9.6/packages/alpinejs/src/directives/x-for.js#L161) because `elForSpot` is undefined. Is this expected behavior? Looking at the diff algorithm in the [`loop`](https://github.com/alpinejs/alpine/blob/v3.9.6/packages/alpinejs/src/directives/x-for.js#L111-L120) function, it looks as though when a previous key is found, the code expects that that's because it existed previously in a rendered table, not that it's a duplicate. 

This [Codepen](https://codepen.io/csfrancis/pen/gOodazp) demonstrates the problem. To work around it, change one of the `time` values.

**Comment by SimoTod** [comment, marked as answer]
Keys must be unique. Even if it didn't throw an exception there, it would break when you reftesh the list.
